### PR TITLE
Added section on Replacing Opcode Handlers

### PR DESCRIPTION
**Pull request by derickr**
As implemented in https://github.com/xdebug/xdebug/pull/559

### CODE REVIEW

File: Book/php7/extensions_design/hooks.rst [hunk 1]
```diff
@@ -350,4 +350,73 @@ overwrite of ``zend_interrupt_function``.
 Replacing Opcode Handlers
 *************************
 
-TODO
+It is possible to override the Zend engine's individual handlers for opcodes. This might be useful
+to ignore the ``@`` operator or count how often each opcode gets executed. The API in the engine
+is only geared up for having *one* extension defined handler per opcode, which makes it important
+that you as extension author, take care of handlers already set by other extensions.
```
Comment (nikic) [inline Book/php7/extensions_design/hooks.rst]:
```suggestion
that you as an extension author, take care of handlers already set by other extensions.
```

File: Book/php7/extensions_design/hooks.rst [hunk 2]
```diff
@@ -350,4 +350,73 @@ overwrite of ``zend_interrupt_function``.
...
+ZEND_USER_OPCODE_CONTINUE
+    Execute next opcode
+ZEND_USER_OPCODE_RETURN
+    Exit from executor (return from function)
+ZEND_USER_OPCODE_DISPATCH
+    Call original opcode handler
+ZEND_USER_OPCODE_ENTER
+    Enter into new op_array without recursion
+ZEND_USER_OPCODE_LEAVE
+    Return to calling op_array within the same executor
```
Comment (nikic) [inline Book/php7/extensions_design/hooks.rst]: I don't think this is going to format properly.

File: Book/php7/extensions_design/hooks.rst [hunk 3]
```diff
@@ -350,4 +350,73 @@ overwrite of ``zend_interrupt_function``.
...
+            return MYEXTG(original_begin_silence_handler)(execute_data);
+        } else {
+            return MYEXTG(original_end_silence_handler)(execute_data);
+        }
+    }
+
+    PHP_MINIT_FUNCTION(my_extension)
+    {
+        MYEXTG(original_begin_silence_handler) = zend_get_user_opcode_handler(ZEND_BEGIN_SILENCE);
+        MYEXTG(original_end_silence_handler) = zend_get_user_opcode_handler(ZEND_END_SILENCE);
```
Comment (nikic) [inline Book/php7/extensions_design/hooks.rst]: Does this return non-null even if no existing handler is set?

File: Book/php7/extensions_design/hooks.rst (hunk 2 above)
Comment (derickr) [reply Book/php7/extensions_design/hooks.rst]: It's called a definition list, and it formats just fine → https://github.com/derickr/PHP-Internals-Book/blob/patch-1/Book/php7/extensions_design/hooks.rst#replacing-opcode-handlers

File: Book/php7/extensions_design/hooks.rst (hunk 3 above)
Comment (derickr) [reply Book/php7/extensions_design/hooks.rst]: It returns NULL if no handler is set, so I've updated the `silence_handler` code to check for this.

File: Book/php7/extensions_design/hooks.rst (hunk 2 above)
Comment (nikic) [reply Book/php7/extensions_design/hooks.rst]: TIL. As you can see, I don't use rst much...

Comment (derickr) [reply Book/php7/extensions_design/hooks.rst]: I do → https://github.com/derickr/derickrethans-articles :-)

**Pull request merged**